### ISSUE TITLE
[Bug]: Recent api not working on macOS

### ISSUE THREAD

**Issue by create3000** [electron/electron#40611]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

27.1.2

### What operating system are you using?

macOS

### Operating System Version

Sonoma

### What arch are you using?

x64

### Last Known Working Electron version

don't know

### Expected Behavior

Recent files added with `app.addRecentDocument` should appear in recentDocuments menu item and in dock app icon menu.

### Actual Behavior

Although I use `app.addRecentDocument` there do not appear any files in recentDocuments menu item, neither in packaged version of app nor in development mode.

**Comment by create3000** [comment]
I created a test project at https://github.com/create3000/electron-recent-documents-api that uses the example code from https://www.electronjs.org/docs/latest/tutorial/recent-documents#example.

To start the program type `npm start` in a terminal.

I tested it also on Windows, but had no luck. I further found out that a recent file is added on macOS under Apple > Recent Objects, but not in the Application menu and not in the Dock menu.

**Comment by caldwell** [comment]
I also just ran into this bug. I checked against some other Electron versions and it appears to be broken in 27.0.0+ (including 28.0.0-beta.11). It works for me in Electron 26.6.2. I am on macOS 14.0 (Sonoma) with an M1 processor.

**Comment by nsfmc** [comment]
i'm able to reproduce this on electron@29.0.0 but as noted, it works fine in the 26.x release.

The menu bar's recents remain dimmed and inactive, showing no submenu. however, right clicking on the app icon does show the recent items there. i looked at the code, namely `Browser::AddRecentDocument` and all i can guess is that `base::apple::FilePathToNSString(path);` might behave differently than `base::mac::FilePathToNSString(path);` https://github.com/electron/electron/blame/995f3dc120281a5e364ff1624ea77ee58ce89947/shell/browser/browser_mac.mm#L163

still, i find that somewhat unlikely given that right clicking on the dock icon _does_ show recent documents, which i assume are populated from the `noteNewRecentDocumentURL` call and the code for `replaceSubmenuShowingRecentDocuments` https://github.com/electron/electron/blob/0b0707145b157343c42266d2586ed9413a1d54f5/shell/browser/ui/cocoa/electron_menu_controller.mm#L264 hasn't been touched in years. i'll look into this more, i haven't tried building electron yet, but those are the places i would probably poke around first.

**Comment by liamcain** [comment]
Also running this issue on electron@28. The recentDocuments are populated correctly in the dock, but the Application Menu "Open Recent" always remains disabled.

**Comment by nsfmc** [comment]
to add more color to this: in electron 29.1.0, when starting a non-prod electron project, items added are registered and available to the dock, but they're not visible as a submenu of the menu bar's "open recent" item. 

When i package the app, however, even in electron 26, while the open recents menu is active, it is entirely populated with empty/nil items.

<img width="315" alt="image" src="https://github.com/electron/electron/assets/216572/158f0ec6-4a08-4097-b139-33fb47d12c0e">

i get nothing in either case, i do however see a handful of `SharedFileList` errors that look like this ```Failed to resolve bookmark 0 for item DC13CF0A-3DEF-4BB2-9440-EB6226BE679B with error: Error Domain=NSOSStatusErrorDomain Code=-43 "fnfErr: File not found" UserInfo={NSLocalizedDescription=BOOKMARK_DATA_AND_URL_ARE_NULL}```

<img width="1363" alt="image" src="https://github.com/electron/electron/assets/216572/122b14f6-fef3-4588-81b1-347d280e0212">

if i run the app using `yarn start` in dev mode in electron 26 (and not higher), i do see the recents populated correctly 

<img width="434" alt="image" src="https://github.com/electron/electron/assets/216572/53867c4c-bfb8-4c07-bc2e-2a9f7f84d296">

as well as in the dock 

<img width="312" alt="image" src="https://github.com/electron/electron/assets/216572/a91ae937-0c9a-48e3-89d1-8e21e9e4ce7a">

i'm running under Sonoma 14.2.1 (23C71), so i'm guessing this entire combination of things has to do with how electron receives the shared file list in newer static builds. pretty confusing but i'll keep poking

**Comment by KevinHughes** [comment]
This is also not working for me in electron 29.3 running macOS Sonoma 14.4.1, either in production (signed and notarized, with a notarized installer) or in non-production debug mode. The "Open Recent" menu stays disabled while recent documents appear in the app's Dock menu. I'm using npm BTW.

**Comment by pedroabreu** [comment]
I'm seeing the same issue on v27+ and managed to narrow it down to a specific nightly version. `v27.0.0-nightly.20230718` is the last working version so `v27.0.0-nightly.20230719` is the first release with the issue present.

Here's the compare between the two releases https://github.com/electron/electron/compare/v27.0.0-nightly.20230718...v27.0.0-nightly.20230719

As for reproduction, on electron fiddle, add this snippet inside the 'app.whenReady'

```js
app.whenReady().then(() => {
  const template = [
  {
    label: "asdasd",
  "submenu":[
    {
      "label":"Open Recent",
      "role":"recentdocuments",
      "submenu":[
        {
          "label":"Clear Recent",
          "role":"clearrecentdocuments"
        }
      ]
    }
  ]
}
]
const menu = Menu.buildFromTemplate(template)

  Menu.setApplicationMenu(menu)
  createWindow()
```

I see some changes in the [electron_menu_controller.mm](https://github.com/electron/electron/issues/40611#diff-0d56377f9cb77917be10876e1f1a63e5dec6655c4d674ebc8f13ead63d0dfa40) related to recentDocuments but this is as far as I've got.

**Comment by pedroabreu** [comment]
This might be fixed in v29, see https://github.com/electron/electron/pull/41978

**Comment by KevinHughes** [comment]
Confirming that this is fixed for me in electron 3.0.4 and works great.

**Comment by jarek-foksa** [comment]
I also confirm this issue has been finally fixed (tested with Electron version 30.0.6).
